extensions/binding/org.eclipse.smarthome.binding.mqtt/src/main/java/org/eclipse/smarthome/binding/mqtt/internal/discovery/NetworkDiscoveryService.java

```diff
@@ -0,0 +1,209 @@
+
+        // Start the connection and wait until timeout or connected callback returns.
+        testConnection.start();
+        lock.tryLock(TIMEOUT_IN_MS, TimeUnit.MILLISECONDS);
+
+        if (testConnection.connectionState() == MqttConnectionState.CONNECTED) {
+            logger.trace("Found service device at {}",
+                    testConnection.getHost() + ":" + String.valueOf(testConnection.getPort()));
+
+            Map<String, Object> properties = new HashMap<>();
+            properties.put("host", testConnection.getHost());
+            properties.put("port", testConnection.getPort());
+
+            ThingUID thingUID = MqttThingID.getThingUID(testConnection.getHost(), testConnection.getPort());
+            thingDiscovered(DiscoveryResultBuilder.create(thingUID).withTTL(120).withProperties(properties)
+                    .withLabel("Mqtt Broker (" + testConnection.getHost() + ")").build());
+        }
+
+        testConnection.stop();
+    }
+
+    /**
+     * Increase a counter and calls stopScan() if counter equals total.
+     *
+     * @param total The total size
+     */
+    synchronized protected void stopScanIfAllScanned(int total) {
+        scannedIPcount += 1;
+        if (scannedIPcount == total) {
+            logger.trace("Scan of {} IPs successful", scannedIPcount);
+            stopScan();
+        }
+    }
+
+    /**
+     * Return a list of IP v4 addresses to be scanned for the discovery
+     */
+    protected List<String> getScannableIPs() {
+        return NetUtil.getAllInterfaceAddresses().stream().filter(i -> i.getAddress() instanceof Inet4Address)
+                .map(interfaceIP -> Arrays
+                        .asList(new SubnetUtils(interfaceIP.getAddress().getHostAddress()).getInfo().getAllAddresses())
+                        .subList(0, MAX_IPS_PER_INTERFACE))
+                .flatMap(List::stream).collect(Collectors.toList());
+    }
+
+    /**
+     * Starts the DiscoveryThread for each IP on each interface on the network
+     */
+    @Override
+    protected void startScan() {
+        removeOlderResults(getTimestampOfLastScan(), null);
+        logger.trace("Starting Discovery");
+
+        List<String> networkIPs = getScannableIPs();
+
+        // We perform two scans per IP (plain/secure)
+        scannedIPcount = 0;
+
+        // Check localhost first
+        networkIPs.add(0, "127.0.0.1");
+        final int totalScans = networkIPs.size() * 2;
+
+        // Create necessary objects to test
+        MqttBrokerConnection o[];
+        try {
+            o = createTestConnections(networkIPs);
+        } catch (ConfigurationException e) {
+            logger.debug("Could not create MqttBrokerConnection object", e);
+            stopScan();
+            return;
+        }
+
+        for (MqttBrokerConnection c : o) {
+            scheduler.execute(() -> {
+                Thread.currentThread().setName("Discovery thread " + c.getHost() + ":" + String.valueOf(c.getPort()));
```

Comment:
Do not change the thread names - they are from a re-used pool and they are already named by the framework.